src/Nethermind/Nethermind.Network/ForkInfo.cs

```diff
@@ -70,5 +85,142 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
+
+                // If our head is beyond this fork, continue to the next (we have a dummy
+                // fork of maxuint64 as the last item to always fail this check eventually).
+                if (i + 1 < Forks.Length && headActivation >= Forks[i + 1].Activation.Activation) continue;
+
+                // Found the first unpassed fork block, check if our current state matches
+                // the remote checksum (rule #1).
+                if (Bytes.AreEqual(forkId.ForkHash, peerId.ForkHash))
+                {
+                    // Fork checksum matched, check if a remote future fork block already passed
+                    // locally without the local node being aware of it (rule #1a).
+                    if (peerId.Next > 0 && headActivation >= peerId.Next)
+                    {
+                        return IForkInfo.ValidationResult.IncompatibleOrStale;
+                    }
+                    // Haven't passed locally a remote-only fork, accept the connection (rule #1b).
+                    return IForkInfo.ValidationResult.Valid;
+                }
+
+                // The local and remote nodes are in different forks currently, check if the
+                // remote checksum is a subset of our local forks (rule #2).
+                for (int j = 0; j < i; j++)
+                {
+                    if (Bytes.AreEqual(Forks[j].Id.ForkHash, peerId.ForkHash))
+                    {
+                        // Remote checksum is a subset, validate based on the announced next fork
+                        if (Forks[j + 1].Activation.Activation != peerId.Next)
+                        {
+                            return IForkInfo.ValidationResult.RemoteStale;
+                        }
+
+                        return IForkInfo.ValidationResult.Valid;
+                    }
+                }
+
+                // Remote chain is not a subset of our local one, check if it's a superset by
+                // any chance, signalling that we're simply out of sync (rule #3).
+                for (int j = i + 1; j < Forks.Length; j++)
+                {
+                    if (Bytes.AreEqual(Forks[j].Id.ForkHash, peerId.ForkHash))
+                    {
+                        // Yay, remote checksum is a superset, ignore upcoming forks
+                        return IForkInfo.ValidationResult.Valid;
+                    }
+                }
+                // No exact, subset or superset match. We are on differing chains, reject.
+                return IForkInfo.ValidationResult.IncompatibleOrStale;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        /// <summary>
+        /// Verify that the forkid from peer matches our forks. Code is largely copied from Geth.
+        /// </summary>
+        /// <param name="peerId"></param>
+        /// <returns></returns>
+        public IForkInfo.ValidationResult ValidateForkId2(ForkId peerId)
```

Comment:
I like 2 a LOT more, avoids linear scan, basically what method 1 does is:
1. Find local forkId based on head with linear scan
2. Compare it with remote forkId, potentially return Valid/IncompatibleOrStale
3. Try find our past fork that is based on peer forkId with linear scan, potentially return Valid/RemoteStale
4. Try find our future fork that is based on peer forkId with linear scan, return Valid/IncompatibleOrStale

Version 2 skips the linear scan, going straight to the matching fork, then checks:
1. If next forks don't match:
2. if this is last remotePeer fork and for us it is not and we activated next one - RemoteStale
3. if peer has next fork and we also activated our fork:
- if we would also activate peerFork -> IncompatibleOrStale
- if we have next fork and we would activate it -> IncompatibleOrStale

This would benefit from better naming, maybe handling all options explicitly?